### DISCUSSION TITLE
How to implment server-side filtering with TreeSelect

### DISCUSSION THREAD

**Discussion by xinghul** [ant-design/ant-design#34924] [General]
### What problem does this feature solve?

I have a large amount of data(300k records) and I can't load them all in the UI, so I need to filter the data and return only about 20 rows for the TreeSelect component.

and the tricky part is I'm doing a fuzzy search on multiple fields on the backend, it's not simply filtering by title.

Right now the searchValue and onSearch only work with client-side search, how do I take a lazy load approach with server-side filtering?

Ideally, if there's a flag to disable searchValue, it would be great, I can load data with onSearch value.

### What does the proposed API look like?

disableSearchValue={true | false}

**Comment by mansion-sun** [comment]
TreeSelect 使用onSearch调用接口查询时，options已经被更新，但是视图中并不会展示最新的options下拉选项，失去焦点后才会展示新的值；这个Select的交互效果不一致
select：
`
      {
        title: "井号",
        dataIndex: "jhdm",
        hideInTable: true,
        valueType: "select",
        formItemProps: {
          rules: [
            {
              required: true,
            },
          ],
        },
        fieldProps: {
          labelInValue: true, // value 类型从 string 变为 { value: string, label: ReactNode } 的格式
          options: wellOptions,
          placeholder: "输入关键字模糊搜索",
          allowClear: true,
          showSearch: true,
          onSearch: debounce((value) => getWellOptions(value, setWellOptions)),
        },
        search: {
          transform: (value) => {
            return {
              jhdm: value?.value,
            };
          },
        },
      },
`
   treeSelect：
`
      {
        title: "井眼名称",
        dataIndex: "czjydm",
        hideInSearch: true,
        hideInTable: true,
        valueType: "treeSelect",
        formItemProps: {
          rules: [
            {
              required: true,
            },
          ],
        },
        fieldProps: {
          allowClear: true,
          labelInValue: true, // value 类型从 string 变为 { value: string, label: ReactNode } 的格式
          options: boreholeOptions,
          placeholder: "输入关键字模糊搜索",
          showSearch: true,
          onSearch: (value) => getBoreholeOptions(value, setBoreholeOptions),
        },
        search: {
          transform: (value) => {
            return {
              czjydm: value?.value,
            };
          },
        },
      },
      `
我尝试了filterTreeNode本地过滤，**treeSelect似乎默认会根据value进行一次过滤**，之后的数据才会在filterTreeNode中二次过滤；这也是接口数据更新之后始终无法回填在treeSelect的options中的原因，**默认的过滤无值，停止了继续更新options**，只要**默认的过滤有值，接口查询到的数据就可以更新到options中，每次接口数据更新之前会展示默认的过滤值**，这真的很无厘头！
      `{
        title: "井眼名称",
        dataIndex: "czjydm",
        hideInSearch: true,
        hideInTable: true,
        valueType: "treeSelect",
        formItemProps: {
          rules: [
            {
              required: true,
            },
          ],
        },
        fieldProps: {
          allowClear: true,
          labelInValue: true, // value 类型从 string 变为 { value: string, label: ReactNode } 的格式
          placeholder: "输入关键字模糊搜索",
          showSearch: true,
        },
        search: {
          transform: (value) => {
            return {
              czjydm: value?.value,
            };
          },
        },
        request: (params, props) => {
          return getBoreholeOptions(params?.keyWords)
        },
      },`
我尝试了Pro中的request异步请求方式，也无法避免默认根据value过滤的问题；
**1.默认过滤有值，会再次更新接口请求的数据为新的options；
2.默认过滤无值，接口请求的数据即便有值也不会更新在视图中为新的options；**

**Reply by allnewers** [reply]
我试了下，把treeNodeFilterProp的值设置成title，应该可以解决。

**Comment by mansion-sun** [comment]
![image](https://user-images.githubusercontent.com/26341016/164168495-f1cd62aa-82ba-49ab-955d-bde479271985.png)
![image](https://user-images.githubusercontent.com/26341016/164168517-99786d57-52a0-4cae-8cd6-ad30fe9d9015.png)

之前开发时没有看到`treeNodeFilterProp | 输入项过滤对应的 treeNode 属性`，treeNodeFilterProp会对TreeSelect远程搜索的接口数据进行二次过滤后更新options，即使接口返回过滤后的数据已经是用户想要的待选项；treeNodeFilterProp能否提供关闭功能，还是必须调用filterTreeNode返回true，人为使客户端过滤失效；
treeNodeFilterProp和searchValue最终似乎都在filterTreeNode中进行调用，在filterTreeNode中关闭客户端过滤更合适，这样可以保证远程过滤不受客户端过滤干扰！
@afc163
